crates/accelerate/src/synthesis/discrete_basis/solovay_kitaev.rs

```diff
@@ -0,0 +1,276 @@
+    }
+
+    /// Run the Solovay Kitaev algorithm, given the initial unitary as U(2) matrix.
+    ///
+    /// This matrix is given using [Complex64] numbers, which can limit the precision of the
+    /// algorithm. It is preferred to run the algorithm using [synthesize_std], which let's Qiskit
+    /// attempt to construct the underlying SO(3) matrix representation using [BigFloat] at
+    /// quadruple precision, instead of inferring it from the double precision U(2) matrix.
+    pub fn synthesize_matrix(
+        &self,
+        matrix_u2: &Matrix2<Complex64>,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_u2(matrix_u2, self.do_checks);
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run the Solovay Kitaev algorithm on a standard gate + gate parameters.
+    ///
+    /// This attempts to directly construct the SO(3) matrix representation to minimize roundoff
+    /// errors. If unsuccessful, this falls back onto constructing the standard U(2) matrix
+    /// representation and converting it to SO(3).
+    pub fn synthesize_std(
+        &self,
+        gate: &StandardGate,
+        params: &[Param],
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_std(gate, params)?;
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run Solovay Kitaev given a [GateSequence].
+    ///
+    /// Mainly an internal helper for other runner methods.
+    fn synthesize_sequence(
+        &self,
+        sequence: &GateSequence,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        // Run SK recursion.
+        let mut output = self.recurse(sequence, recursion_degree);
+
+        // Do minor optimizations on the output sequence, e.g. cancel gate-inverse-pairs, which
+        // can arise due to the recursions (even though they should not be present in the initial
+        // set of basic approximations).
+        output.inverse_cancellation();
+
+        // Compute the error in the SO(3) representation to return with the circuit.
+        let circuit = output.to_circuit(Some(sequence))?;
+        Ok(circuit)
+    }
+
+    /// Run a recursion step for a gate sequence, given a recursion degree.
+    ///
+    /// If the degree is 0 (recursion root), return the closest approximation in the set of
+    /// basic approximations. Otherwise, decompose the difference between the approximation
+    /// and the target unitary as balanced group commutator, and recurse on each element.
+    fn recurse(&self, sequence: &GateSequence, degree: usize) -> GateSequence {
+        // Recursion root: return the best approximation in the precomputed set.
+        if degree == 0 {
+            if self.do_checks {
+                math::assert_so3("Recursion root", &sequence.matrix_so3);
+            }
```

Comment:
These checks are `false` by default, right? Would it make sense to add pure Rust tests (in case this is not done already)?